File: CameraFilter/src/main/java/com/frank/camerafilter/util/OpenGLUtil.java
package com.frank.camerafilter.util;

import android.content.Context;
import android.content.res.AssetManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.opengl.GLES11Ext;
import android.opengl.GLES30;
import android.opengl.GLUtils;
import android.util.Log;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

import javax.microedition.khronos.opengles.GL10;

public class OpenGLUtil {

    public final static int ON_DRAWN   = 1;
    public static final int NOT_INIT   = -1;
    public static final int NO_SHADER  = 0;
    public static final int NO_TEXTURE = -1;

    private static Bitmap getBitmapFromAssetFile(Context context, String name) {
        try {
            AssetManager assetManager = context.getResources().getAssets();
            InputStream stream = assetManager.open(name);
            Bitmap bitmap = BitmapFactory.decodeStream(stream);
            stream.close();
            return bitmap;
        } catch (IOException e) {
            return null;
        }
    }

    public static int loadTexture(final Context context, final String name) {
        if (context == null || name == null)
            return NO_TEXTURE;
        final int[] textures = new int[1];
        GLES30.glGenTextures(1, textures, 0);
        if (textures[0] == 0)
            return NO_TEXTURE;
        Bitmap bitmap = getBitmapFromAssetFile(context, name);
        if (bitmap == null)
            return NO_TEXTURE;
        GLES30.glBindTexture(GLES30.GL_TEXTURE_2D, textures[0]);
        GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_MIN_FILTER, GLES30.GL_LINEAR);
        GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_MAG_FILTER, GLES30.GL_LINEAR);
        GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_WRAP_S,     GLES30.GL_CLAMP_TO_EDGE);
        GLES30.glTexParameteri(GLES30.GL_TEXTURE_2D, GLES30.GL_TEXTURE_WRAP_T,     GLES30.GL_CLAMP_TO_EDGE);
        GLUtils.texImage2D(GLES30.GL_TEXTURE_2D, 0, bitmap, 0);
        bitmap.recycle();
        return textures[0];
    }

    private static int loadShader(final String source, final int type) {
        int shader = GLES30.glCreateShader(type);
        GLES30.glShaderSource(shader, source);
        GLES30.glCompileShader(shader);
        int[] compile = new int[1];
        GLES30.glGetShaderiv(shader, GLES30.GL_COMPILE_STATUS, compile, 0);
        if (compile[0] <= 0) {
            Log.e("OpenGlUtil", "Shader compile error=" + GLES30.glGetShaderInfoLog(shader));
            return NO_SHADER;
        }
        return shader;
    }

    public static int loadProgram(final String vertexSource, final String fragmentSource) {
        int vertexShader   = loadShader(vertexSource, GLES30.GL_VERTEX_SHADER);
        int fragmentShader = loadShader(fragmentSource, GLES30.GL_FRAGMENT_SHADER);
        if (vertexShader == NO_SHADER || fragmentShader == NO_SHADER) {
            return 0;
        }
        int programId = GLES30.glCreateProgram();
        GLES30.glAttachShader(programId, vertexShader);
        GLES30.glAttachShader(programId, fragmentShader);
        GLES30.glLinkProgram(programId);
        int[] linked = new int[1];
        GLES30.glGetProgramiv(programId, GLES30.GL_LINK_STATUS, linked, 0);
        if (linked[0] <= 0) {
            programId = 0;
            Log.e("OpenGlUtil", "program link error=" + GLES30.glGetProgramInfoLog(programId));
        }
        GLES30.glDeleteShader(vertexShader);
        GLES30.glDeleteShader(fragmentShader);
        return programId;
    }

    public static int getExternalOESTextureId() {
        int[] textures = new int[1];
        GLES30.glGenTextures(1, textures, 0);
        GLES30.glBindTexture(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, textures[0]);
        GLES30.glTexParameterf(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, GL10.GL_TEXTURE_MIN_FILTER, GL10.GL_LINEAR);
        GLES30.glTexParameterf(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, GL10.GL_TEXTURE_MAG_FILTER, GL10.GL_LINEAR);
        GLES30.glTexParameteri(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, GL10.GL_TEXTURE_WRAP_S, GL10.GL_CLAMP_TO_EDGE);
        GLES30.glTexParameteri(GLES11Ext.GL_TEXTURE_EXTERNAL_OES, GL10.GL_TEXTURE_WRAP_T, GL10.GL_CLAMP_TO_EDGE);
        return textures[0];
    }

    public static String readShaderFromSource(Context context, final int resourceId) {
        String line;
        StringBuilder builder = new StringBuilder();
        InputStream inputStream = context.getResources().openRawResource(resourceId);
        InputStreamReader reader = new InputStreamReader(inputStream);
        BufferedReader bufferedReader = new BufferedReader(reader);
        try {
            while ((line = bufferedReader.readLine()) != null) {
                builder.append(line).append("\n");
            }
        } catch (IOException e) {
            return null;
        } finally {
            try {
                inputStream.close();
                reader.close();
                bufferedReader.close();
            } catch (IOException e) {
                e.printStackTrace();
            }

        }
        return builder.toString();
    }

}

File: engine/render/opengl/common/OpenGLUtil.cpp
/**
 * Note: operation of OpenGL
 * Date: 2025/12/15
 * Author: frank
 */

#include "OpenGLUtil.h"

#include "NextLog.h"

#define OPENGL_TAG "OpenGLUtil"

GLuint CreateProgram(const std::string &vertexShader, const std::string &fragmentShader) {
    if (vertexShader.empty()) {
        NEXT_LOGE(OPENGL_TAG, "vertex shader is empty!\n");
        return -1;
    }
    if (vertexShader.empty() || fragmentShader.empty()) {
        NEXT_LOGE(OPENGL_TAG, "fragment shader is empty!\n");
        return -1;
    }

    GLint compileStatus = 0;
    GLint linkStatus = GL_FALSE;
    const char *vertexShaderStr   = vertexShader.c_str();
    const char *fragmentShaderStr = fragmentShader.c_str();

    GLuint program    = glCreateProgram();
    GLuint vertShader = glCreateShader(GL_VERTEX_SHADER);
    GLuint fragShader = glCreateShader(GL_FRAGMENT_SHADER);
    if (!vertShader) {
        NEXT_LOGE(OPENGL_TAG, "create vertex shader error, code=%d\n", glGetError());
        return -1;
    }
    if (!fragShader) {
        NEXT_LOGE(OPENGL_TAG, "create fragment shader error, code=%d\n", glGetError());
        goto error;
    }

    glShaderSource(vertShader, 1, &vertexShaderStr, nullptr);
    glCompileShader(vertShader);
    glGetShaderiv(vertShader, GL_COMPILE_STATUS, &compileStatus);
    if (!compileStatus) {
        NEXT_LOGE(OPENGL_TAG, "compile vertex shader error, source=%s\n", vertexShaderStr);
        goto error;
    }

    glShaderSource(fragShader, 1, &fragmentShaderStr, nullptr);
    glCompileShader(fragShader);
    glGetShaderiv(fragShader, GL_COMPILE_STATUS, &compileStatus);
    if (!compileStatus) {
        NEXT_LOGE(OPENGL_TAG, "compile fragment shader error, source=%s\n", fragmentShaderStr);
        goto error;
    }

    glAttachShader(program, vertShader);
    glAttachShader(program, fragShader);
    glLinkProgram(program);
    glGetProgramiv(program, GL_LINK_STATUS, &linkStatus);
    if (!linkStatus) {
        NEXT_LOGE(OPENGL_TAG, "link program error, code=%d\n", glGetError());
        goto error;
    }

error:
    if (vertShader) {
        glDeleteShader(vertShader);
    }
    if (fragShader) {
        glDeleteShader(fragShader);
    }

    return program;
}

GLuint CreateTexture() {
    auto *texture = new GLuint[1];
    glGenTextures(1, texture);
    glBindTexture(GL_TEXTURE_2D, texture[0]);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
    glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
    glTexParameterf(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
    glBindTexture(GL_TEXTURE_2D, 0);
    GLuint result = texture[0];
    delete[] texture;

    return result;
}

GLuint CreateFramebuffer(int width, int height, GLuint texture) {
    auto *frameBuffer = new GLuint[1];
    glGenFramebuffers(1, frameBuffer);
    glBindFramebuffer(GL_FRAMEBUFFER, frameBuffer[0]);
    glBindTexture(GL_TEXTURE_2D, texture);
    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, nullptr);
    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, texture, 0);
    glBindTexture(GL_TEXTURE_2D, 0);
    glBindFramebuffer(GL_FRAMEBUFFER, 0);
    GLuint result = frameBuffer[0];
    delete[] frameBuffer;

    return result;
}
